### PR TITLE
Igor/setting embedded

### PR DESCRIPTION
**Pull request by igorT**

### CODE REVIEW

**Comment by igorT** [comment]
@hjdivad I wasn't sure what the best location for the tests was, so kept the original one. Happy to move them

File: addon/record-data.js [hunk 1]
```diff
@@ -861,9 +860,11 @@ export default class M3RecordData {
...
   }
 
-  _setChildRecordData(key, idx, record) {
-    let recordData = recordDataFor(record);
-
+  _setChildRecordData(key, idx, recordData) {
+    if (recordData._baseRecordData && this._baseRecordData) {
+      this._baseRecordData._setChildRecordData(key, idx, recordData._baseRecordData);
+    }
+    // TODO assert against one of these being set but the other one not
```
Comment (hjdivad) [inline addon/record-data.js]: add this assert?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 2]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+        }
+
+        computeNestedModel(key, value) {
+          assert.ok(true, 'Called computed nested model');
+          let attributesType;
+          try {
+            attributesType = value && value.$type;
+          } catch (e) {
+            assert.ok(false, "We don't pass Megamorphic Models to computeNestedModel");
+          }
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: what's the point of `try/catch` -> assert.  We'll get a failure if we let this exception go through no?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 3]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+
+          if (key !== 'results' && typeof attributesType !== 'string') {
+            // object without a known type
+            return undefined;
+          }
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: does this check have anything to do with the test?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 4]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+          type: 'com.bookstore.Book',
+          attributes: {
+            locations: [
+              {
+                country: 'US',
+                geographicArea: 'California',
+                city: 'San Francisco',
+                postalCode: '94110',
+                description: 'Club house',
+                $type: 'com.linkedin.voyager.organization.OrganizationAddress',
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: let's change this `$type` to something more generic

File: tests/unit/model/projections/changed-attrs-test.js [hunk 5]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+                headquarter: true,
+                line1: '1234 Lucky St',
+              },
+            ],
+          },
+        },
+      ],
+    });
+
+    let model = this.store.peekRecord('com.bookstore.Book', 'urn:book:1');
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: is the correct terminology not record?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 6]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+
+    const currentCollection = model.get('locations').slice();
+    const aNewLocation = {
+      country: 'MX',
+      geographicArea: 'California',
+      city: 'Ensenada',
+      postalCode: '22810',
+      description: 'Home',
+      $type: 'com.linkedin.voyager.organization.OrganizationAddress',
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: type -> generic type

File: tests/unit/model/projections/changed-attrs-test.js [hunk 7]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
       'changed attributes is correct'
     );
   });
+
+  test('Can set a many embedded property to a semi resolved array containing a mix of pojos and megamorphic models', function(assert) {
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: let's move this test out of projections; something like `tests/unit/model/changed-attrs` or `tests/unit/model/nested-model`.

I'm not too concerned as long as it's not wrong (eg non-proj test in proj) and as long as we're not creating giant test files (eg `model-test.js`)

File: tests/unit/model/projections/changed-attrs-test.js [hunk 8]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+      postalCode: '22810',
+      description: 'Home',
+      $type: 'com.linkedin.voyager.organization.OrganizationAddress',
+      headquarter: true,
+      line1: '555 Main St.',
+    };
+    model.set('locations', currentCollection.concat(aNewLocation));
+
+    let locations = model.get('locations');
+    assert.ok(locations, 'Locations retrieved succesfully');
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: let's add an assertion that `locations` is as we expect

File: tests/unit/model/projections/changed-attrs-test.js [hunk 9]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+        }
+
+        computeNestedModel(key, value) {
+          assert.ok(true, 'Called computed nested model');
+          let attributesType;
+          try {
+            attributesType = value && value.$type;
+          } catch (e) {
+            assert.ok(false, "We don't pass Megamorphic Models to computeNestedModel");
+          }
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: do we need this try/catch?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 10]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+
+          if (key !== 'results' && typeof attributesType !== 'string') {
+            // object without a known type
+            return undefined;
+          }
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: what does this conditional get us?  at a minimum, i don't think `results` is a special key in this test

File: tests/unit/model/projections/changed-attrs-test.js [hunk 11]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+          type: 'com.bookstore.Book',
+          attributes: {
+            locations: [
+              {
+                country: 'US',
+                geographicArea: 'California',
+                city: 'San Francisco',
+                postalCode: '94110',
+                description: 'Club house',
+                $type: 'com.linkedin.voyager.organization.OrganizationAddress',
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: type -> generic type

File: tests/unit/model/projections/changed-attrs-test.js [hunk 12]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+          id: 'urn:book:1',
+          type: 'com.bookstore.ProjectedBook',
+          locations: [
+            {
+              country: 'US',
+              geographicArea: 'California',
+              city: 'San Francisco',
+              postalCode: '94110',
+              description: 'Club house',
+              $type: 'com.linkedin.voyager.organization.OrganizationAddress',
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: type -> generic type

File: tests/unit/model/projections/changed-attrs-test.js [hunk 13]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+              headquarter: true,
+              line1: '1234 Lucky St',
+            },
+          ],
+        },
+      ],
+    });
+
+    let model = this.store.peekRecord('com.bookstore.ProjectedBook', 'urn:book:1');
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: model -> record no?

File: tests/unit/model/projections/changed-attrs-test.js [hunk 14]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+
+    const currentCollection = model.get('locations').slice();
+    const aNewLocation = {
+      country: 'MX',
+      geographicArea: 'California',
+      city: 'Ensenada',
+      postalCode: '22810',
+      description: 'Home',
+      $type: 'com.linkedin.voyager.organization.OrganizationAddress',
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: type -> generic type

File: tests/unit/model/projections/changed-attrs-test.js [hunk 15]
```diff
@@ -88,4 +88,191 @@ module('unit/model/projections/changed-attrs', function(hooks) {
...
+      headquarter: true,
+      line1: '555 Main St.',
+    };
+
+    model.set('locations', currentCollection.concat(aNewLocation));
+    // Saving relies on correctly setup nested record datas
+    await model.save();
```
Comment (hjdivad) [inline tests/unit/model/projections/changed-attrs-test.js]: shouldn't we assert something after the save?

Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: We are making sure calling save doesn't error, I can put an assert in the mocked adapter to make sure it's called

File: tests/unit/model/projections/changed-attrs-test.js (hunk 3 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: It mimics the failure mode discovered, I think we still need at least a part of this code to get it to fail robustly, but I can try and narrow down which ones.

File: addon/record-data.js (hunk 1 above)
Comment (igorT) [reply addon/record-data.js]: done

File: tests/unit/model/projections/changed-attrs-test.js (hunk 2 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: fixed

File: tests/unit/model/projections/changed-attrs-test.js (hunk 4 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: fixed

File: tests/unit/model/projections/changed-attrs-test.js (hunk 5 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: fixed

File: tests/unit/model/projections/changed-attrs-test.js (hunk 6 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: fixed

File: tests/unit/model/projections/changed-attrs-test.js (hunk 3 above)
Comment (igorT) [reply tests/unit/model/projections/changed-attrs-test.js]: Fixed

**Pull request merged**